### ISSUE TITLE
Determine how to handle saturation for the e-image files

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#29]
On DC1 scale time frames we are using e-image files with only saturation.  We are assuming everything else is corrected as DM does not yet have tools to handle all of the other sensor effects.

There is also the related issue of non-linearity but I think that should show up in the electronics module.  Saturation is really an electron effect in the sensor.

There are (at least?) three (?) ways saturation can be included:

- (Best) The new framework that @rmjarvis is preparing in GalSim-developers/GalSim#799 could include by hosting the new sensor models which include it.   Currently the models are really focused on BF and don't yet implement this.

- In the shorter term can Mike's framework apply this at one of the steps?

- Is there an option to already to this in the photon shooting in GalSim?  I looked at issues etc and I don't see anything.  But, somehow I thought I remembered seeing this option before being used in the context of WFIRST.

I think we need this before we can do meaningful comparisons with PhoSim .  Can @rmjarvis or @rmandelb comment on they best way to do this?

**Comment by tonytyson** [comment]
Craig's model now includes saturation and associated charge skimming.

**Comment by rmandelb** [comment]
If Craig's model now includes a validated model for saturation in addition to b/f, I think that's clearly the best option.

Otherwise I think it would not be hard to include in photon shooting, but this is not currently done.  It was briefly considered as something we'd implement along the nonlinearity in https://github.com/GalSim-developers/GalSim/issues/552 but then the decision was made to have nonlinearity stand on its own and you're right, there is no open issue for it.

**Comment by rmjarvis** [comment]
I'm not sure that the sensor step is the right place for this.  At least if you want proper treatment of bleeding, which spreads charge through the whole length of the CCD.  The sensor application is currently envisioned to be used for single objects, so it is just run on postage stamps.

To be sure, there are some issues to be dealt with about how best to deal with overlapping objects and parts of the image where there are no objects (i.e. only draw sky photons), but I'm not sure that we want to have the sensor step trying to bleed charge along the whole length of the readout rows.  

It seems to me that it would be better to do this once you have the final image with how many electrons are intended to go into each pixel and overflow as necessary at that point.  i.e. at the readout stage.  @jchiang87 has implemented something to deal with many of the amplifier effects.  The saturation bit (and nonlinearities and probably other things) would go before this, but in the same slot, so the readout stage would a list of processes to apply, not (necessarily) just one step.

**Comment by dkirkby** [comment]
Craig's code does exhibit saturation, but this occurs at much higher charges than observed in data so I wouldn't call it "validated".  It is possible that there are important dynamic effects associated with the readout charge transfer that set the practical saturation threshold, but these are beyond the scope of Craig's electrostatic model (unless there are updates that I don't know about and which are not on github).

I agree with @rmjarvis that the readout and associated effects should be treated as a separate step, after the pseudo-static sensor response step that provides the total charge in each pixel.

**Comment by tonytyson** [comment]
We find in the lab that saturation sets in gradually and can be seen in 
cores of stars even at 50,000e/pixel.  It is due to the space charge 
effect moving the potential well and thus charges closer to the 
frontside surface where there are states in the gap. This creates 
missing charge and a flat-top PSF profile.  Distinct from BF effect.

**Comment by dkirkby** [comment]
The Poisson simulations predict that a weak buried channel is still present after adding ~400Ke, so the electrostatic saturation effects do not seem to describe the lab data well.  This is illustrated in the plot below (from my talk at the Oxford SAWG session):

![saturation](https://cloud.githubusercontent.com/assets/185007/19865889/53539824-9f5b-11e6-91bb-36512fb67472.png)

The dotted curve shows the buried channel before any charge is present, and the solid curve shows the potential after added 410Ke.  (The histogram shows the charge density but on a different scale - it should be peaked as the same place as the potential).

**Comment by cwwalter** [comment]
> The sensor application is currently envisioned to be used for single objects, so it is just run on postage stamps.

Just as an aside I didn't quite follow this comment... I thought we build up the true e-image picture and then pass a complete list of photons and their positions to the sensor model?  That's why it seemed to me saturation was most naturally handled in that sensor model. FWIW I recall that is the way it works in PhoSim.

**Comment by rmjarvis** [comment]
> I thought we build up the true e-image picture and then pass a complete list of photons and their positions to the sensor model? 

We can discuss this as an option, but that's not currently the way I was imagining it working.  Mostly because I think the memory demands of keeping all of the photons around for the whole image before actually raining them down through the sensor would be too severe. 

But let's discuss in person next week what the trade offs are for the different ways of doing this.  I think the recent changes I made to the mid-level API should give us the flexibility to do it whichever way we want.

**Comment by cwwalter** [comment]
> We can discuss this as an option, but that's not currently the way I was imagining it working. Mostly because I think the memory demands of keeping all of the photons around for the whole image before actually raining them down through the sensor would be too severe.
> 
> But let's discuss in person next week what the trade offs are for the different ways of doing this. I think the recent changes I made to the mid-level API should give us the flexibility to do it whichever way we want.

Unfortunately I can't come next week due to teaching and other travel etc..  I'm afraid if things don't all come sampled randomly spatially at the same time than BF won't work.  Were you thinking of another way to make sure that was OK?

You could bin everything first and then draw from the histogram but that would lose any frequency/angle info. You could draw from those too...

**Comment by rmjarvis** [comment]
It's certainly easier to get all the bookkeeping related to blends and nearby objects right if we just generate the full image first and then run that through the sensor model.  The sky too is easier to deal with in that framework.  So maybe that's a better way to go.

The main downside is that we'll lose the sub-pixel locations of the photons for the case where we use photon shooting to build the e-image.  But that might not be that big a deal -- especially if we oversample the intermediate image by a factor of 2x2.  Then Lanczos interpolation to draw the photon positions would probably be fine.

Maybe worth trying to figure out how much memory overhead would be required to save all the original photons for a full CCD.  If it's not too huge, we could try to get the best of both worlds by saving the sensor step to the end, but still use all the original photons.

**Comment by rmjarvis** [comment]
I should add that in GalSim, photons are allowed to have a flux != 1.  So a single "photon" can represent more than one physical photon.  For very bright objects, this is typically how we make the photon shooting method not take forever.

This means that the storage required for keeping all the photons won't get overwhelmed by a few bright stars.  They can have chunky photons that each carry flux >> 1.

**Comment by rmandelb** [comment]
Does that optimization potentially stop working once we want to include certain detector effects?  (saturation / blooming / brighter-fatter / insert your favorite effect that screws up the distribution of photons compared to what it would have been for a low-flux star)

**Comment by rmjarvis** [comment]
I don't think so.  I think most applications are fine with doing several photons at a time between updates.  I know Craig has talked about dropping something like 1000 photons between updates to the pixel edges for B/F.  So in the case of photons with flux > 1, that just means that you go until the flux of the "photons" you added exceed 1000, rather than just counting 1000 photons.

**Comment by cwwalter** [comment]
> Maybe worth trying to figure out how much memory overhead would be required to save all the original photons for a full CCD. If it's not too huge, we could try to get the best of both worlds by saving the sensor step to the end, but still use all the original photons.

There are many good reasons for keeping all of the true photon information.  One use case to think about is related to blending.  I started to do some work in PhoSim and would like to come back to it here where the true object ID was associated with the converted electrons.  This way, you can know the truth information when you are trying to test the deblending algorithms.  There are several ways to approach this.  I was making  a 3d fits cube with each layer corresponding to an object.  I talked to @gpdf and David about ways to visualize this in the SUIT.  

This probably doesn't scale very well to 250K objects but we could think about other ways of associating the information.  We may want to make another issue somewhere to address this particular idea, but I wanted to mention how generally powerful I think having access to this information could be for studies.

**Comment by cwwalter** [comment]
How this is being done has now been determined @craiglagegit has implemented saturation properly in his Poisson simulator after adding realistic doping profiles from a sensor post-mortem.  We still need to validate the GalSim version of this and implement bleeding in the electronics readout but those issues are handled in #7 and #6.  

Closing this issue.
